modules/gdscript/gdscript_analyzer.cpp

```diff
@@ -783,7 +783,16 @@ void GDScriptAnalyzer::resolve_class_member(GDScriptParser::ClassNode *p_class,
 					}
 					datatype = member.variable->initializer->get_datatype();
 
-					if (datatype.type_source != GDScriptParser::DataType::UNDETECTED) {
+					if (member.variable->infer_datatype) {
+						if (!datatype.is_set() || datatype.has_no_type()) {
+							push_error(vformat(R"(Cannot infer the type of "%s" variable because the initial value doesn't have a set type.)", member.variable->identifier->name), member.variable->initializer);
+						} else if (datatype.is_variant() && !datatype.is_hard_type()) {
+							push_error(vformat(R"(Cannot infer the type of "%s" variable because the initial value is Variant. Use explicit "Variant" type if this is intended.)", member.variable->identifier->name), member.variable->initializer);
+						} else if (datatype.kind == GDScriptParser::DataType::BUILTIN && datatype.builtin_type == Variant::NIL) {
+							push_error(vformat(R"(Cannot infer the type of "%s" variable because the initial value is "null".)", member.variable->identifier->name), member.variable->initializer);
+						}
+						datatype.type_source = GDScriptParser::DataType::ANNOTATED_INFERRED;
```

Comment:
I wonder whether this should be set here, given that errors may occur. I don't know what the behavior of the analyzer is elsewhere if it the type source is `ANNOTATED_INFERRED` but the datatype isn't set.

Should it be set like this only for when there are no errors?

Reply:
Good notice. Yes, it is always set to `ANNOTATED_INFERRED` (in variables, constants and params) ignoring errors. Fix for it should not be hard - new variable `bool is_inferred = false`, additional last `else` branch after errors that set `is_inferred = true` and then `type_source` is set in ternary to either  `ANNOTATED_INFERRED` or `UNDETECTED` based on the bool.

This should be done in other PR for all those places.

Reply:
I was thinking about this a little more and wondering whether this should change or not. Do you think this is actually an issue?

I've noticed that the analyzer is very permissive - it seems to prefer annotating the first error, and ignoring any errors stemming from the first error. This way it's more obvious which is the _main_ or _original_ error. Do you think that's why this currently works this way?

Reply:
Yeah, actually came to the same conclusion after some time. In the end it is a preference and I think that less noise is better, since you seem to agree I'll resolve the conversion.